rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,283 @@
+
+```nix
+problemHandler = {
+  "*" = {
+    "*" = "throw";
+    alias = "trace";
+    maintainerless = "ignore";
+  };
+  myPackage.foo = "ignore";
+};
+```
+
+If multiple rules match a given problem of a package, the most specific handler will be called:
+
+1. `pkgName.problemName`
+2. `pkgName.problemKind`
+3. `pkgName."*"`
+4. `"*".problemName`
+5. `"*".problemKind`
+6. `"*"."*"`
+
+So for example the value of `myPackage.*` would override the one in `*.maintainerless` if both matched, as it is more specific.
+
+The default value in nixpkgs might be something like:
+
+```nix
+problemHandler."*" = {
+  "*" = "trace";
+  removal = "throw";
+  deprecated = "throw";
+  maintainerless = "ignore";
+};
+```
+
+It may also be expanded by values from other configuration options as part of a migration scheme from the other mechanisms.
+
+## Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+### Propagation across transitive dependencies
+
+When a package has a problem that `throw`s, all packages that depend on it will fail to evaluate until that problem is ignored or resolved. Most of the time, this is sufficient.
+
+When the problem requires actions on dependents however, it does not sufficiently inform about all packages that need action. Multiple packages may be annotated with the same problem, in that case it should be given a name and the name should be the same across all instances. Other values like the message or the URL list do not need to be the same and may be adapted sensibly.
+
+For the example of Python 2 deprecation, all problems would have `name = "python2-eol"` and then a user may set `config."*".python2-eol = "ignore";` to ignore them.
+
+### Backwards compatbility, Backporting and stable releases
+
+New problems generally should not be added to stable branches if possible, and also not be backported to them, since it may break evaluation. The same rule applies to other changes to a pacakge's `meta` which may generate a problem and thus lead to evaluation failure too. Scenarios where evaluation failure is a desired goal, for example with unfixable security issues, are obviously exempt from this.
+
+### Removal of packages
+
+The plan is to eventually remove packages with long outstanding problems. The details will be part of future work, but at the very least a package must have a problem whose kind defaults to "throw" for at least one full release cycle (so that stable users have sufficient time to be warned and intervene).
+
+If a package needs to be removed for some other reason, the problem kind `removal` should be used instead:
+
+```nix
+meta.problems = [{
+  kind = "removal";
+  message = "We don't want this in nixpkgs anymore";
+  date = "1970-01-01";
+}];
+```
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+- Too much warnigns may cause Alarm fatigue
```

Comment:
When I was deciding between gentoo and nixpkgs, this was the major strike against gentoo.  The OS *sends you email*.  And you sort of have to read it, because all the other users do.

That, plus wanting less Python in my life rather than more, is what tipped the balance.

Let's not go that way.

Reply:
Currently, nixpkgs errs on the other side of the spectrum. It is difficult for you to get notified about changes you care about, especially on the unstable channel. This RFC has started because people wanted to do some backwards-incompatible changes but did not have a good way to properly inform all stakeholders.